driver-kotlin-sync/src/test/kotlin/com/mongodb/kotlin/client/MongoClientTest.kt

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.kotlin.client
+
+import com.mongodb.ClientSessionOptions
+import com.mongodb.client.MongoClient as JMongoClient
+import kotlin.reflect.full.declaredFunctions
+import kotlin.test.assertEquals
+import org.bson.BsonDocument
+import org.bson.Document
+import org.junit.jupiter.api.Test
+import org.mockito.Mock
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.refEq
+import org.mockito.kotlin.times
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.verifyNoMoreInteractions
+import org.mockito.kotlin.whenever
+
+class MongoClientTest {
+
+    @Mock val wrapped: JMongoClient = mock()
+    @Mock val clientSession: ClientSession = ClientSession(mock())
+
+    @Test
+    fun shouldHaveTheSameMethods() {
+        val jMongoClientFunctions = JMongoClient::class.declaredFunctions.map { it.name }.toSet()
+        val kMongoClientFunctions = MongoClient::class.declaredFunctions.map { it.name }.toSet()
+
+        assertEquals(jMongoClientFunctions, kMongoClientFunctions)
+    }
+
+    @Test
+    fun shouldCallTheUnderlyingClose() {
+        val mongoClient = MongoClient(wrapped)
+        mongoClient.close()
+
+        verify(wrapped).close()
+        verifyNoMoreInteractions(wrapped)
+    }
+
+    @Test
+    fun shouldCallTheUnderlyingClusterDescription() {
+        val mongoClient = MongoClient(wrapped)
+        whenever(wrapped.clusterDescription).doReturn(mock())
+
+        mongoClient.getClusterDescription()
+
+        verify(wrapped).clusterDescription
```

Comment:
These wrap tests take a different approach (one per method, individual mocks, etc.) compared to some of the others (where many methods are tested in a row, even lower down in this class). Why the difference?

Reply:
So we seem to have a couple of conventions:

1) A test for each method. Which applies to the MongoClient, Database and Collection. Where there are lots of methods and overloads for each method.
2) Usage based tests case are more brief and tied to usage. Which includes chainable API's like MongoIterable and then an outlier with MongoCursor.

This follows on from the Scala convention, which probably followed on from the convention established when the reactive stream driver wrapped the async driver. The original reasons for such an approach is lost.

Reply:
Alright